<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [56]:
import pandas as pd
import psycopg2

In [57]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [58]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [59]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [60]:
# текст запроса
query_3_1 = f'''select count (id) 
            from public.vacancies
             '''

In [61]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [62]:
# текст запроса
query_3_2 = f'''select count (id) 
            from public.employers
             '''

In [63]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [64]:
# текст запроса
query_3_3 = f'''select count (id) 
            from public.areas
             '''

In [65]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [66]:
# текст запроса
query_3_4 = f'''select count (id) 
            from public.industries
             '''

In [67]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# Выводы по предварительному анализу данных
По результатам предварительного анализа данных установлено что всего представлено:
    49179 вакансий;
    23501 работодатель;
    1362 региона;
    294 сферы деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [68]:
# текст запроса
query_4_1 = f'''select a.name, count (v.id)
            from public.vacancies v
            left join public.areas a on a.id = v.area_id
            group by a.name
            order by count (v.id) desc
            limit 5
          '''

In [69]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [70]:
# текст запроса
query_4_2 = f'''select count (v.id)
            from public.vacancies v
            where salary_from is not null or salary_to is not null
            '''

In [71]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [72]:
# текст запроса
query_4_3 = f'''select round (avg (v.salary_from)), round (avg (v.salary_to))
            from public.vacancies v
             '''

In [73]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [74]:
# текст запроса
query_4_4 = f'''select v.schedule, v.employment, count (v.id)
            from public.vacancies v
            group by v.schedule, v.employment
            order by count (v.id) desc
            offset 1  --убираем 1
            limit 1 --Выводим 1 (т.е 2 после того как убрали 1)
             '''

In [75]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [76]:
# текст запроса
query_4_5 = f'''select v.experience, count (v.id)
            from public.vacancies v
            group by v.experience
            order by count (v.id) asc
            '''

In [77]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# Выводы по детальному анализу вакансий
Детальный анализ вакансий показал, что большинство вакансий расположены в городе Москве, минимальная средняя заработная плата составляет 71000 рублей, максимальная средняя 110000 рублей, предпочтительный тип рабочего графика и трудоустройства на первом месте (Полный день - полная занятость) на втором месте (Удаленная работа-Полная занятость), болинство работников требуются с опытом работы от года до 3.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [78]:
# текст запроса
query_5_1 = f'''select e.name, count (v.id)
            from public.employers e
            left join public.vacancies v on e.id = v.employer_id
            group by e.name
            order by count (v.id) desc
            limit 5
            '''

In [79]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [80]:
# текст запроса
query_5_2 = f'''select a.name, count (v.name), count (e.id)
            from public.areas a
            left join public.vacancies v on a.id = v.area_id
            left join public.employers e on a.id = e.area
            group by a.name
            having count (v.name) = 0
            order by count (e.id) desc
            limit 1          
            '''

In [81]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count  count
0  Россия      0    410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [82]:
# текст запроса
query_5_3 = f'''select e.name, count (distinct (a.name))
            from public.vacancies v
            join public.employers e on v.employer_id = e.id
            join public.areas a on v.area_id = a.id
            group by e.name
            order by count (a.name) desc
            limit 1         
            '''

In [83]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0  Яндекс    181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [84]:
# текст запроса
query_5_4 = f'''select count (e.name), count (ei.industry_id)
            from public.employers e
            left join public.employers_industries ei on ei.employer_id = e.id
            where ei.industry_id is null        
            '''

In [85]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count  count
0   8419      0

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [86]:
# текст запроса
query_5_5 = f'''select e.name, count (i.name)
            from public.employers_industries ei
            left join public.employers e on ei.employer_id = e.id
            left join public.industries i on ei.industry_id = i.id
            group by e.name
            having count (i.name) = 4
            order by e.name asc
            offset 2
            limit 1       
            '''

In [87]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [88]:
# текст запроса
query_5_6 = f'''select count (e.name), i.name
            from public.employers_industries ei
            left join public.employers e on ei.employer_id = e.id
            left join public.industries i on ei.industry_id = i.id
            where i.name = 'Разработка программного обеспечения'
            group by i.name       
            '''

In [89]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count                                 name
0   3553  Разработка программного обеспечения

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [90]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup 
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
tbody = page.find ('tbody') # тег в котором хранится вся информация из таблицы
a = tbody.find_all ('a') # тег с названиями городов
cities = [] # Создаем список
for city in a: # Перебираем все теги и получам текст из них
    cities.append (city.text) # Добавляем текст в список
cities = cities[5:] # Убираем лишнии данные из тегов
print (cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [91]:
# текст запроса
query_5_7 = f'''(select a.name, count (e.name) 
            from public.vacancies v
            left join public.employers e on v.employer_id = e.id
            left join public.areas a on v.area_id = a.id
            where a.name in  ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск',
            'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград') and e.name = 'Яндекс' 
            group by a.name
            having count (e.name) > 0
            order by count (e.name) asc)
            union all
            (select 'total', count (e.name)
            from public.vacancies v
            left join public.employers e on v.employer_id = e.id
            left join public.areas a on v.area_id = a.id
            where a.name in  ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск',
            'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград') and e.name = 'Яндекс')
            '''


In [92]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  count
0             total    485
1              Омск     21
2         Челябинск     22
3        Красноярск     23
4         Волгоград     24
5             Пермь     25
6            Казань     25
7    Ростов-на-Дону     25
8               Уфа     26
9            Самара     26
10        Краснодар     30
11          Воронеж     32
12      Новосибирск     35
13  Нижний Новгород     36
14     Екатеринбург     39
15  Санкт-Петербург     42
16           Москва     54

***

# Выводы по анализу работадателей
В ходе детального анализа установлено, что на территории России представлено наибольшее количество работадателей, работадателем у когорого наибольшее количество вакансий является "Яндекс". Вакансии "Яндекса" размещены в 181 регионе из них 16 регионов города миллионники, количество обьявлений в этих городах составило 485. Наибольшее количество предсталенных вакансий связаны с разработкой программного обеспечния.  

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [93]:
# текст запроса
query_6_1 = f'''select count (v.name) 
            from public.vacancies v
            where v.name ILIKE '%data%' or v.name ILIKE '%данн%' --ILIKE невосприимчив к регистрам      
            '''

In [94]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [95]:
# текст запроса
query_6_2 = f'''select count (v.name) 
            from public.vacancies v
            WHERE
            (
            (v.name ILIKE '%data scientist%') OR (v.name ILIKE '%data science%') OR (v.name ILIKE '%исследователь данных%')
            OR ((v.name ILIKE '%ML%') AND (v.name NOT ILIKE '%HTML%'))
            OR (v.name ILIKE '%machine learning%')
            OR (v.name ILIKE '%машинн%обучен%')
            AND (v.key_skills IS NOT NULL)
            )
            AND (v.name ILIKE '%junior%'
            OR v.experience ILIKE '%Нет опыта%'
            OR v.employment ILIKE '%Стажировка%')      
            '''


In [96]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [97]:
# текст запроса
query_6_3 = f'''select count(v.name)
            from public.vacancies v
            WHERE
            (
            (v.name ILIKE '%data scientist%' OR v.name ILIKE '%data science%' OR v.name ILIKE '%исследователь данных%'
            OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
            OR v.name ILIKE '%machine learning%'
            OR v.name ILIKE '%машинн%обучен%'))
            and (v.key_skills ILIKE '%sql%' or v.key_skills ILIKE '%postgres%')       
            '''

In [98]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [99]:
# текст запроса
query_6_4 = f'''select count(v.name)
            from public.vacancies v
            WHERE
            (
            (v.name ILIKE '%data scientist%' OR v.name ILIKE '%data science%' OR v.name ILIKE '%исследователь данных%'
            OR (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
            OR v.name ILIKE '%machine learning%'
            OR v.name ILIKE '%машинн%обучен%'))
            and (v.key_skills ILIKE '%python%')       
            '''

In [100]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [101]:
# текст запроса
query_6_5 = f'''select
            round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2)
            from    
            vacancies
            where
            (( name ILIKE '%data scientist%' OR 
            name ILIKE '%data science%' OR 
            name ILIKE '%исследователь данных%' OR 
            name ILIKE '%machine learning%' OR 
            name ILIKE '%машинн%обучен%' OR 
            name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
            and key_skills is not null)
            '''

In [102]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [103]:
# текст запроса
query_6_6 = f'''select v.experience, 
            round(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)
            from public.vacancies v
            WHERE (
            (v.name ILIKE '%data scientist%')
            OR (lower(v.name) ILIKE '%data science%')
            OR (v.name ILIKE '%исследователь данных%')
            OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
            OR (v.name ILIKE '%machine learning%')
            OR (v.name ILIKE '%машинн%обучен%')
            AND (key_skills IS NOT NULL)
            )
            GROUP BY v.experience
            ORDER BY 1
            '''

In [104]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  145236.0
3       От 3 до 6 лет  243115.0

***

# выводы по предметному анализу
В ходе предметного анализа, проанализированы вакансии связанные с работой с данными. Из полученных результатов можно для работы дата саентинстом необходимо обладать следующими знаниями:
1. Владеть SQL и postgres
2. Знать язык программирования Python
3. Для конкурентно способной заработной платы необходимо обладать опытом от 3 до 6 лет.

# Общий вывод по проекту

# Общие выводы по проекту
Анализ вакансий НН показал, что большинство вакансий расположены в городе Москве,  предпочтительный тип рабочего графика и трудоустройства (Полный день - полная занятость).
Работадателем у когорого наибольшее количество вакансий является "Яндекс". Вакансии "Яндекса" размещены в 181 регионе из них 16 регионов города миллионники, количество обьявлений в этих городах составило 485. Наибольшее количество предсталенных вакансий связаны с разработкой программного обеспечния.
В части поиска работы DataScientist необходимо обладать компетенциями изложенными в выводах по предметному анализу и опыт работы от 3 до 6 (желательно для наибольшей ЗП).

# Дополнительное исследование (В каких 5 городах требуется большее количество DataScientis)

In [105]:
# текст запроса
query_dop = f'''select a.name, count (v.name) 
            from public.vacancies v
            left join public.employers e on v.employer_id = e.id
            left join public.areas a on v.area_id = a.id
            WHERE ((v.name ILIKE '%data scientist%')
            OR (v.name ILIKE '%data science%')
            OR (v.name ILIKE '%исследователь данных%')
            OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
            OR (v.name ILIKE '%machine learning%')
            OR (v.name ILIKE '%машинн%обучен%')
            AND (key_skills IS NOT NULL))
            GROUP BY a.name
            ORDER by count (v.name) desc
            limit 5
            '''

In [106]:
# результат запроса
df = pd.read_sql_query(query_dop, connection)
df

/var/folders/nx/6955y2w53pz25qw4tzbs7vpw0000gn/T/ipykernel_41508/2727121611.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_dop, connection)


name  count
0           Москва    216
1  Санкт-Петербург     64
2      Новосибирск     23
3  Нижний Новгород     20
4           Казань     16

# Выводы по дополнительным исследованиям
DataScientistу будет проще всего найти работу в городе Москва, наименее сложно в таких городах как Санкт-Петербург, Новосибирск, Нижний Новгород, Казань (городах Миллионниках). 

In [107]:
connection.close()